File: app/src/main/java/co/smartreceipts/android/model/impl/columns/categories/CategoryColumnDefinitions.java
package co.smartreceipts.android.model.impl.columns.categories;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.StringRes;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import co.smartreceipts.android.R;
import co.smartreceipts.android.model.Column;
import co.smartreceipts.android.model.ColumnDefinitions;
import co.smartreceipts.android.model.comparators.ColumnNameComparator;
import co.smartreceipts.android.model.impl.columns.AbstractColumnImpl;
import co.smartreceipts.android.model.impl.columns.BlankColumn;
import co.smartreceipts.android.persistence.database.controllers.grouping.results.SumCategoryGroupingResult;
import co.smartreceipts.android.sync.model.SyncState;
import co.smartreceipts.android.sync.model.impl.DefaultSyncState;
import wb.android.flex.Flex;

public class CategoryColumnDefinitions implements ColumnDefinitions<SumCategoryGroupingResult> {

    private enum ActualDefinition {
        NAME(R.string.category_name_field),
        CODE(R.string.category_code_field),
        PRICE(R.string.category_price_field),
        TAX(R.string.category_tax_field),
        CURRENCY(R.string.category_currency_field);

        private final int stringResId;

        ActualDefinition(int stringResId) {
            this.stringResId = stringResId;
        }

        @StringRes
        public final int getStringResId() {
            return stringResId;
        }

    }

    private final ActualDefinition[] actualDefinitions;
    private final Flex flex;
    private final Context context;


    public CategoryColumnDefinitions(Flex flex, Context context) {
        this.flex = flex;
        this.context = context;
        this.actualDefinitions = ActualDefinition.values();
    }

    @Override
    public Column<SumCategoryGroupingResult> getColumn(int id, @NonNull String definitionName, @NonNull SyncState syncState) {
        for (int i = 0; i < actualDefinitions.length; i++) {
            final ActualDefinition definition = actualDefinitions[i];
            if (definitionName.equals(getColumnNameFromStringResId(definition.getStringResId()))) {
                return getColumnFromClass(id, definition, definitionName, syncState);
            }
        }
        return null;
    }

    @NonNull
    @Override
    public List<Column<SumCategoryGroupingResult>> getAllColumns() {
        final ArrayList<AbstractColumnImpl<SumCategoryGroupingResult>> columns = new ArrayList<>(actualDefinitions.length);
        for (int i = 0; i < actualDefinitions.length; i++) {
            final ActualDefinition definition = actualDefinitions[i];
            final AbstractColumnImpl<SumCategoryGroupingResult> column = getColumnFromClass(Column.UNKNOWN_ID, definition, getColumnNameFromStringResId(definition.getStringResId()), new DefaultSyncState());
            if (column != null) {
                columns.add(column);
            }
        }
        Collections.sort(columns, new ColumnNameComparator<AbstractColumnImpl<SumCategoryGroupingResult>>());
        return new ArrayList<Column<SumCategoryGroupingResult>>(columns);
    }

    @NonNull
    @Override
    public Column<SumCategoryGroupingResult> getDefaultInsertColumn() {
        return new BlankColumn<>(Column.UNKNOWN_ID, context.getString(R.string.column_item_blank), new DefaultSyncState());
    }


    private AbstractColumnImpl<SumCategoryGroupingResult> getColumnFromClass(int id, @NonNull ActualDefinition definition, @NonNull String definitionName, @NonNull SyncState syncState) {
        switch (definition) {
            case NAME:
                return new CategoryNameColumn(id, definitionName, syncState);
            case CODE:
                return new CategoryCodeColumn(id, definitionName, syncState);
            case PRICE:
                return new CategoryPriceColumn(id, definitionName, syncState);
            case TAX:
                return new CategoryTaxColumn(id, definitionName, syncState);
            case CURRENCY:
                return new CategoryCurrencyColumn(id, definitionName, syncState);
            default:
                throw new IllegalArgumentException("Unknown definition type: " + definition);
        }
    }

    private String getColumnNameFromStringResId(int stringResId) {
        if (flex != null) {
            return flex.getString(context, stringResId);
        } else {
            return context.getString(stringResId);
        }
    }
}

File: app/src/main/java/co/smartreceipts/android/model/impl/columns/categories/CategoryColumnDefinitions.kt
package co.smartreceipts.android.model.impl.columns.categories

import androidx.annotation.StringRes
import co.smartreceipts.android.R
import co.smartreceipts.android.model.ActualColumnDefinition
import co.smartreceipts.android.model.Column
import co.smartreceipts.android.model.ColumnDefinitions
import co.smartreceipts.android.model.Keyed
import co.smartreceipts.android.model.comparators.ColumnNameComparator
import co.smartreceipts.android.model.impl.columns.AbstractColumnImpl
import co.smartreceipts.android.model.impl.columns.categories.CategoryColumnDefinitions.ActualDefinition.*
import co.smartreceipts.android.persistence.database.controllers.grouping.results.SumCategoryGroupingResult
import co.smartreceipts.core.sync.model.SyncState
import co.smartreceipts.core.sync.model.impl.DefaultSyncState
import co.smartreceipts.android.workers.reports.ReportResourcesManager
import java.util.*

class CategoryColumnDefinitions(private val reportResourcesManager: ReportResourcesManager, private val multiCurrency: Boolean) :
    ColumnDefinitions<SumCategoryGroupingResult> {

    private val actualDefinitions = values()

    /**
     * Note: Column types must be unique
     * Column type must be >= 0
     */
    internal enum class ActualDefinition(
        override val columnType: Int,
        @StringRes override val columnHeaderId: Int
    ) : ActualColumnDefinition {
        NAME(0, R.string.category_name_field),
        CODE(1, R.string.category_code_field),
        PRICE(2, R.string.category_price_field),
        TAX(3, R.string.category_tax_field),
        PRICE_EXCHANGED(4, R.string.category_price_exchanged_field);

    }

    override fun getColumn(
        id: Int,
        columnType: Int,
        syncState: SyncState,
        ignoredCustomOrderId: Long,
        ignoredUuid: UUID
    ): Column<SumCategoryGroupingResult> {
        for (definition in actualDefinitions) {

            if (columnType == definition.columnType) {
                return getColumnFromClass(definition, id, syncState)
            }
        }
        throw IllegalArgumentException("Unknown column type: $columnType")
    }

    override fun getAllColumns(): List<Column<SumCategoryGroupingResult>> {
        val columns = ArrayList<AbstractColumnImpl<SumCategoryGroupingResult>>(actualDefinitions.size)

        for (definition in actualDefinitions) {
            // don't include PRICE_EXCHANGED definition if all receipts have same currency
            if (!(definition == PRICE_EXCHANGED && !multiCurrency)) {
                columns.add(getColumnFromClass(definition))
            }
        }

        Collections.sort(columns, ColumnNameComparator(reportResourcesManager))
        return ArrayList<Column<SumCategoryGroupingResult>>(columns)
    }

    override fun getDefaultInsertColumn(): Column<SumCategoryGroupingResult> {
        return getColumnFromClass(NAME)
    }


    private fun getColumnFromClass(
        definition: ActualDefinition,
        id: Int = Keyed.MISSING_ID,
        syncState: SyncState = DefaultSyncState()
    ): AbstractColumnImpl<SumCategoryGroupingResult> {
        return when (definition) {
            NAME -> CategoryNameColumn(id, syncState)
            CODE -> CategoryCodeColumn(id, syncState)
            PRICE -> CategoryPriceColumn(id, syncState)
            TAX -> CategoryTaxColumn(id, syncState)
            PRICE_EXCHANGED -> CategoryExchangedPriceColumn(id, syncState)
        }
    }
}
